### ISSUE TITLE
@WithMockOAuth2Authentication doesn't work due to a security fix from spring

### ISSUE THREAD

**Issue by xetys** [jhipster/generator-jhipster#4058]
Well, the next situation is quite complex, and needs some decision to go on

During my recent work I invented the above mentioned annotation in order to provide some tool to integration test security configuration, which is a very important feature for me, since it's a must have if you are TDD style... 

There also was a question, wether to PR it to spring, and here is some problem: That annotation, in the way I contributed it to JHipster, won't work at all. The reason why it worked to me in July, I simply forgot. It is the stateless flag in spring `OAuth2AuthenticationProcessingFilter`, which clears the security context, when it is present. This is done to fix some whole, where you can get accidentaly authenticated used some cookies (and I have no idea, how this is working, it is described in incode comments)...in other words, only real tokens lead to real authentication....the `@WithSecurityContext` approach is cutted away by spring itself

But, there ***would*** be a way to make it working: 

```  java

    @Override
    public void configure(ResourceServerSecurityConfigurer resources) throws Exception {
        if (Arrays.asList(environment.getActiveProfiles()).contains("test")) {
            resources.stateless(false);
        }
    }
```

in `MicroserviceSecurityConfiguration`. As of @deepu105 's opinion, this is mixing testing stuff into actual implementation. Disabling stateless mode at all, is a no way, as I will never sleep peacefully anymore if know, I was contributing some insecure hack in order to satisfy some needs for testing...

I'm already looking for a cleaner way to solve this, but at the moment, that is the only way I know which is working...so I need your opinion on this issue :smile:

**Comment by deepu105** [comment]
I would rather skip a test case than opening a security hole :) but thats just my opinion.

**Comment by xetys** [comment]
actually, this would only mean to maintain some hacky code, in your to satisfy both needs.

for the security hole thing I am currently discussing on gitter with Rob and Dave....as it currently seems the actual security considarion still exists, if no cookie based authenticated happen in that application....as we don't use cookies for microservice applications. I think if there is some security problem, then if using classic stateful security and oauth support in one application.
If this holds, spring could just add a configuration property, to make `private boolean stateless = true` in `OAuth2AuthenticationProcessingFilter` be configurable to some option, having default value "true".

Then, it would be just com extra config in applications.yml in test package, what would be totally clean. I am currently trying to argue this in https://gitter.im/spring-projects/spring-security

I will report

**Comment by xetys** [comment]
Ok, to sum up.....they don't think it is a good way to make that stateless flag configurable, as my needs of testability can be achieved using `@MockBean`

## **tl;dr;** Version

That annotation must be rewritten in a way, it uses `@MockBean` on `ResourceServerTokenServices` to say "if token XXX gets in, grant the following authentication" and this won't be cleared out by spring as it is the correct way of OAuth2. Furthermore the restMockMvc should add an autorization header to that request......all this baked up using that annotation, or some library function call instead (which I do not prefer, but it looks like a giant task for me to make such annotation)

## more detailed version

Here is my [experimental code][], where I use the `@MockBean` thing in order to mock a authentication into spring security oauth...and it works, but not the elegant way, since it consists of 2 parts, mocking and header setting in the acutal call...I need more annotation wizzardy to solve this or some Util lib instead


[experimental code]: https://github.com/xetys/jhipster-uaa-setup/blob/full-uaa-setup/app1/src/test/java/com/mycompany/myapp/web/rest/SecurityIntTest.java

**Comment by xetys** [comment]
to solve this problem in a way which satisfies all side I choosed to use a different approach, in short:

instead of using the `@WithMockOAuth2Authentication` annotation, the developer must do 2 things

1. inject OAuth2TokenMockUtil to the test, were mocking should happen

2. do this:

``` java

    @Test
    public void testUnauthorizedAccess() throws Exception {
        restMockMvc.perform(get("/api/foos")
            .with(oAuth2TokenMockUtil.oauth2Authentication("test")))
            .andExpect(status().isForbidden());
    }

    @Test
    public void testAuthorizedAccess() throws Exception {
        restMockMvc.perform(get("/api/foos")
            .with(oAuth2TokenMockUtil.oauth2Authentication("test", Sets.newSet("web-app"))))
            .andExpect(status().isOk());
    }
```
***(assuming applySecurity() is active and "/api/foo/" restricted to scope "web-app")***

This still is clean in my point of view, even without fancy annotation, and is no longer opening security holds, as it mocks some internal bean responsible for token encoding instead of providing a `SecurityContext`, which is usually cleared out because of the stateless flag.

I will directly provide update on documentation as soon this gets merged

**Comment by jannis-baratheon** [comment]
Leaving this post here for reference. I spent quite a while trying to tacle this. Maybe I'll save someone's time in the future.

I wanted to implement security access tests in my monolithic JHipster app. I couldn't get `OAuth2TokenMockUtil` to work, I kept on getting an error like:

    expected single matching bean but found 3: consumerTokenServices,defaultAuthorizationServerTokenServices,org.springframework.security.oauth2.provider.token.ResourceServerTokenServices#0

FYI The only thing that worked for me wa thes 'mocked stateless flag' approach (found it [on StackOverflow](https://stackoverflow.com/a/31641775/4494577) prior to getting here).

Snippets:

* OAuth2TestConfiguration.java (in `src/test/`)

		@Configuration
		public class OAuth2TestConfiguration {

			@Bean
			@Primary
			public Boolean oauth2StatelessSecurityContext() {
				return Boolean.FALSE;
			}
		}

* OAuth2ServerConfiguration (standard JHipster-generated OAuth2 configuration - only differences highlighted)

		@Configuration
		public class OAuth2ServerConfiguration {
		 
			(...) 
		 
			@Configuration
			@EnableResourceServer
			protected static class ResourceServerConfiguration extends ResourceServerConfigurerAdapter {
		
				@Autowired(required = false)
				@Qualifier("oauth2StatelessSecurityContext")
				private Boolean stateless = Boolean.TRUE; // STATEFUL switching for tests!
	   
				(...)
	   
				@Override
				public void configure(ResourceServerSecurityConfigurer resources) throws Exception {
					resources
						.stateless(this.stateless)
						.resourceId("res_testapp")
						.tokenStore(tokenStore);
				}
				
				(...)
			}
			
			(...)
		}

* The test (in src/test)

		@RunWith(SpringRunner.class)
		@SpringBootTest(classes = TestApp.class)
		@ContextConfiguration
		@WebAppConfiguration
		public class TestEntityResourceSecurityIntTest {

			@Autowired
			private WebApplicationContext context;

			private MockMvc mvc;

			@Before
			public void setup() {
				mvc = MockMvcBuilders
					.webAppContextSetup(context)
					.apply(springSecurity())
					.build();
			}

			@Test
			@WithMockUser(authorities = AuthoritiesConstants.ADMIN)
			public void testAdminResourceOk() throws Exception {
				mvc
					.perform(
						get("/api/test-entities")
					)
					.andExpect(status().isOk());
			}

			@Test
			@WithMockUser(authorities = AuthoritiesConstants.USER)
			public void testAdminResourceForbidden() throws Exception {
				mvc
					.perform(
						get("/api/test-entities")
					)
					.andExpect(status().isForbidden());
			}

			@Test
			public void testAdminResourceUnsatuhorized() throws Exception {
				mvc
					.perform(
						get("/api/test-entities")
					)
					.andExpect(status().isUnauthorized());
			}
		}

* Endpoint - TestEntityController.java

		@RestController
		@RequestMapping("/api")
		@Secured("ROLE_ADMIN")
		public class TestEntityResource {

			(...)
			
			@GetMapping("/test-entities")
			@Timed
			public ResponseEntity<List<TestEntityDTO>> getAllTestEntities(@ApiParam Pageable pageable) {
				Page<TestEntityDTO> page = testEntityService.findAll(pageable);
				HttpHeaders headers = PaginationUtil.generatePaginationHttpHeaders(page, "/api/test-entities");
				return new ResponseEntity<>(page.getContent(), headers, HttpStatus.OK);
			}
			
			(...)

		}

**Comment by xetys** [comment]
I think the issue here is the `@MockBean`, as it might have a different behavior of picking the proper bean, in contrast to `@Autowired`. Maybe [this](https://stackoverflow.com/questions/39431630/mockbean-annotation-in-spring-boot-test-causes-nouniquebeandefinitionexception) helps you? 

These utils were built for the microservice OAuth solution. It might should also work for monoliths OAuth solution if there is any interest to this.
